src/Rings/oscar_singular.jl

```diff
@@ -0,0 +1,458 @@
+              _variables_for_singular(symbols(Rx)),
+              cached = false)[1]
+  end
+  return OscarSingularPolyRingMap(Rx, Sx, fcoeff)
+end
+
+function iso_oscar_singular_poly_ring(Rx::MPolyRing, ord::Symbol)
+  fcoeff = iso_oscar_singular_coeff_ring(base_ring(Rx))
+  S = codomain(fcoeff)
+  Sx =  Singular.polynomial_ring(S,
+              _variables_for_singular(symbols(Rx)),
+              ordering = ord,
+              cached = false)[1]
+  return OscarSingularPolyRingMap(Rx, Sx, fcoeff)
+end
+
+function iso_oscar_singular_poly_ring(Rx::MPolyRing, ord::Singular.sordering)
+  fcoeff = iso_oscar_singular_coeff_ring(base_ring(Rx))
+  S = codomain(fcoeff)
+  Sx =  Singular.polynomial_ring(S,
+              _variables_for_singular(symbols(Rx)),
+              ordering = ord,
+              cached = false)[1]
+  return OscarSingularPolyRingMap(Rx, Sx, fcoeff)
+end
+
+function iso_oscar_singular_poly_ring(Rx::MPolyRing, ord::MonomialOrdering)
+  fcoeff = iso_oscar_singular_coeff_ring(base_ring(Rx))
+  S = codomain(fcoeff)
+  Sx =  Singular.polynomial_ring(S,
+              _variables_for_singular(symbols(Rx)),
+              ordering = singular(ord),
+              cached = false)[1]
+  return OscarSingularPolyRingMap(Rx, Sx, fcoeff)
+end
+
+function image(f::OscarSingularPolyRingMap, a)
+  parent(a) !== domain(f) && error("Element not in domain")
+  g = MPolyBuildCtx(codomain(f))
+  for (c, e) = zip(Nemo.coefficients(a), Nemo.exponent_vectors(a))
+    push_term!(g, f.f(c), e)
+  end
+  return finish(g)
+end
+
+function preimage(f::OscarSingularPolyRingMap, a; check = true)
+  check && (parent(a) === codomain(f) || error("Element not in codomain"))
+  g = MPolyBuildCtx(domain(f))
+  for (c, e) = Base.Iterators.zip(AbstractAlgebra.coefficients(a), AbstractAlgebra.exponent_vectors(a))
+    push_term!(g, preimage(f.f, c), e)
+  end
+  return finish(g)
+end
+
+# Quotient rings
+
+struct OscarSingularPolyRingQuoMap{U, V, W} <: Map{U, V, Any, Any}
+  R::U
+  S::V
+  f::W #= isomorphism of the underlying "base rings" =#
+end
+
+domain(f::OscarSingularPolyRingQuoMap) = f.R
+
+codomain(f::OscarSingularPolyRingQuoMap) = f.S
+
+function (f::OscarSingularPolyRingQuoMap)(a)
+  return image(f, a)
+end
+
+function _iso_oscar_singular_poly_ring(R::MPolyQuoRing)
+  _groebner_basis(R)
+  Rorig = base_ring(R)
+  f = iso_oscar_singular_poly_ring(Rorig)
+  @assert base_ring(codomain(f)) === base_ring(R.SQR)
+  return OscarSingularPolyRingQuoMap(R, R.SQR, f)
+end
+
+function image(f::OscarSingularPolyRingQuoMap, b::MPolyQuoRingElem)
+  @assert parent(b) === domain(f)
+  a = b.f
+  return image(f, a)
+end
+
+# for some reason this is used
+function image(f::OscarSingularPolyRingQuoMap, b::MPolyRingElem)
+  @assert parent(b) === base_ring(domain(f))
+  return codomain(f)(b)
+end
+
+function preimage(f::OscarSingularPolyRingQuoMap, a::Singular.spoly)
+  @assert parent(a) === codomain(f)
+  return MPolyQuoRingElem(preimage(f.f, a; check = false), domain(f))
```

Comment:
Does this also work? Then it would seem preferable to me (avoiding the explicit type `MPolyQuoRingElem`)
```suggestion
  return domain(f)(preimage(f.f, a; check = false))
```